call.go

```diff
@@ -29,14 +29,27 @@ import (
 func (cc *ClientConn) Invoke(ctx context.Context, method string, args, reply interface{}, opts ...CallOption) error {
 	// allow interceptor to see all applicable call options, which means those
 	// configured as defaults from dial option as well as per-call options
-	opts = append(cc.dopts.callOptions, opts...)
+	opts = combine(cc.dopts.callOptions, opts)
 
 	if cc.dopts.unaryInt != nil {
 		return cc.dopts.unaryInt(ctx, method, args, reply, cc, invoke, opts...)
 	}
 	return invoke(ctx, method, args, reply, cc, opts...)
 }
 
+func combine(o1 []CallOption, o2 []CallOption) []CallOption {
+	// we don't use append because o1 could have extra capacity whose
+	// elements would be overwritten, which could cause inadvertent
+	// sharing (and race connditions) between concurrent calls
+	if len(o1) == 0 && len(o2) == 0 {
+		return nil
+	}
+	ret := make([]CallOption, 0, len(o1)+len(o2))
+	ret = append(ret, o1...)
```

Comment:
Does this `append` method have the same performance as

```go
ret := make([]CallOption, len(o1)+len(o2))
copy(ret, o1)
copy(ret[len(o1):], o2)
return ret
```

?

Reply:
I went ahead and changed it.

But I was also curious about the performance, so I benchmarked it. With the following benchmark, they are indistinguishable from each other:

```go
func combineViaCopy(o1 []CallOption, o2 []CallOption) []CallOption {
	// we don't use append because o1 could have extra capacity whose
	// elements would be overwritten, which could cause inadvertent
	// sharing (and race connditions) between concurrent calls
	if len(o1) == 0 {
		return o2
	} else if len(o2) == 0 {
		return o1
	}
	ret := make([]CallOption, len(o1)+len(o2))
	copy(ret, o1)
	copy(ret[len(o1):], o2)
	return ret
}

func combineViaAppend(o1 []CallOption, o2 []CallOption) []CallOption {
	// we don't use append because o1 could have extra capacity whose
	// elements would be overwritten, which could cause inadvertent
	// sharing (and race connditions) between concurrent calls
	if len(o1) == 0 {
		return o2
	} else if len(o2) == 0 {
		return o1
	}
	ret := make([]CallOption, 0, len(o1)+len(o2))
	ret = append(ret, o1...)
	ret = append(ret, o2...)
	return ret
}

var co5 = []CallOption{Header(nil), Trailer(nil), Peer(nil),  FailFast(false), MaxCallRecvMsgSize(100)}
var co10 = []CallOption{
	Header(nil), Trailer(nil), Peer(nil),  FailFast(false), MaxCallRecvMsgSize(100),
	Header(nil), Trailer(nil), Peer(nil),  FailFast(false), MaxCallRecvMsgSize(100),
}

func BenchmarkCopy5_5(b *testing.B){
	for i := 0; i < b.N; i++ {
		combineViaCopy(co5, co5)
	}
}

func BenchmarkCopy5_10(b *testing.B){
	for i := 0; i < b.N; i++ {
		combineViaCopy(co5, co10)
	}
}

func BenchmarkAppend5_5(b *testing.B){
	for i := 0; i < b.N; i++ {
		combineViaAppend(co5, co5)
	}
}

func BenchmarkAppend5_10(b *testing.B){
	for i := 0; i < b.N; i++ {
		combineViaAppend(co5, co10)
	}
}
```

On my 2016 model 15" Macbook Pro (2.2Ghz Intel Core i7, 1.6Ghz DDR3 RAM):

```
BenchmarkCopy5_5-8                	10000000	       127 ns/op
BenchmarkCopy5_10-8               	10000000	       160 ns/op
BenchmarkAppend5_5-8              	10000000	       129 ns/op
BenchmarkAppend5_10-8             	10000000	       159 ns/op
```

Reply:
Good to know, thanks for testing!